[Security Solution] Extend the `/upgrade/_review` endpoint contract and functionality

jpdjere opened the pull request:
**Resolves: https://github.com/elastic/kibana/issues/180153**
**Resolves: https://github.com/elastic/kibana/issues/188277**

## Summary

- Extend the POST /upgrade/_review API endpoint's contract and functionality 
- Changes `has_conflict` property within each rule field's `ThreeWayDiff` from `boolean` to `enum` with possible values:
    - `NONE`: no conflicts in three way diff
    - `SOLVABLE`: conflict detected but was successfully resolved by our algorithms
    - `NON_SOLVABLE`: conflict detected and could not be resolved by our algorithms. 

- Adds `has_base_version` boolean field within each field diff calculation. Has values:
    - true: the base version of the field was found and is either defined or undefined
    - false: the base version of the field was not found

- The possible values for `has_conflict` for each concrete diff algorithm are:
    - **single line strings**: `NO`, `NON_SOLVABLE` 
    - **multi line strings**: `NO`, `SOLVABLE`, `NON_SOLVABLE` 
    - **numbers**: `NO`, `NON_SOLVABLE` 
    - **array of scalar values**: `NO`, `SOLVABLE`

- [ ] Adds new logic to handle https://github.com/elastic/kibana/pull/186435#issuecomment-2178484224


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

jpdjere committed:
Create ConflictResolution enum

jpdjere committed:
Update prop name

jpdjere committed:
Fixes

jpdjere committed:
Endpoint updates

jpdjere commented:
/ci

jpdjere committed:
Replaced enum

jpdjere commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/220383)
* Commit: 7afd8975058375d103b204c8c1bab0725200437b

### Failed CI Steps
* [Jest Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909803-6c67-4e71-b73f-8f78ba32a0f4)
* [Jest Tests #15](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909803-6c6a-4c73-8245-a814e3d6ce83)
* [Jest Tests #15](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909826-f145-47f1-92b2-6da7b53f20e6)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909803-6c6a-4c73-8245-a814e3d6ce83) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220383/jobs/01909803-6c6a-4c73-8245-a814e3d6ce83/artifacts/01909826-d834-4276-a412-676ed90473a9) Jest Tests #<span></span>15 / scalarArrayDiffAlgorithm edge cases compares arrays deduplicated when values are duplicated in all versions
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909826-f145-47f1-92b2-6da7b53f20e6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220383/jobs/01909826-f145-47f1-92b2-6da7b53f20e6/artifacts/01909837-9149-4b3c-846c-d145d2a88474) Jest Tests #<span></span>15 / scalarArrayDiffAlgorithm edge cases compares arrays deduplicated when values are duplicated in all versions
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909803-6c6a-4c73-8245-a814e3d6ce83) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220383/jobs/01909803-6c6a-4c73-8245-a814e3d6ce83/artifacts/01909826-d835-4e01-a8fc-a9f131ad720a) Jest Tests #<span></span>15 / scalarArrayDiffAlgorithm returns custom merged version as merged output if all three versions are different - scenario ABC
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909826-f145-47f1-92b2-6da7b53f20e6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220383/jobs/01909826-f145-47f1-92b2-6da7b53f20e6/artifacts/01909837-9148-42f1-8592-a258b347a302) Jest Tests #<span></span>15 / scalarArrayDiffAlgorithm returns custom merged version as merged output if all three versions are different - scenario ABC
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220383#01909803-6c67-4e71-b73f-8f78ba32a0f4) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220383/jobs/01909803-6c67-4e71-b73f-8f78ba32a0f4/artifacts/0190982c-a102-4691-9c9e-8f6c75a1ea08) Jest Tests #<span></span>4 / UnifiedFieldList <FieldNameSearch /> should render correctly

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":220383}
buildkite-pr-comment-->

jpdjere committed:
Fix tests

jpdjere commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/220620)
* Commit: b83a88c70440bd629e76951ddf7e4a71a1d2950e

### Failed CI Steps
* [Defend Workflows Cypress Tests on Serverless #7](https://buildkite.com/elastic/kibana-pull-request/builds/220620#01909cce-fb15-41a6-8171-06da99dfc57b)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/220620#01909cce-fb15-41a6-8171-06da99dfc57b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/220620/jobs/01909cce-fb15-41a6-8171-06da99dfc57b/artifacts/01909d02-097e-45b2-af54-ebbd6544f5c0) Defend Workflows Cypress Tests on Serverless #<span></span>7 / Endpoints page "before all" hook for "Shows endpoint on the list" "before all" hook for "Shows endpoint on the list"

### History
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":220620}
buildkite-pr-comment-->

jpdjere committed:
Added stats

jpdjere committed:
Extend tests

jpdjere committed:
fix type

jpdjere commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

dplumlee commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff.ts:
```diff
@@ -19,6 +19,12 @@ import type { ThreeWayMergeOutcome } from './three_way_merge_outcome';
 export type MissingVersion = typeof MissingVersion;
 
+export enum ThreeWayDiffConflictResolutionResult {
+  NO = 'NO',
```
Should this be something like `NO_CONFLICT`? I guess `NO` makes sense in the `conflict: no` assignment but it does look a bit weird standalone

jpdjere replied:
Updated the enum for clarity

dplumlee reviewed (commented):
This looks good for the most part @jpdjere! Just left one comment about naming and we can figure out how to sync with https://github.com/elastic/kibana/pull/188022

jpdjere committed:
Add upgrade review endpoint stats integration tests

jpdjere committed:
Test name

jpdjere committed:
Updated ENUM

jpdjere committed:
Updated comment

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222026)
* Commit: 3dba3538ee1aded3c384bed3475fb56096a1ebea

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/222026#0190c065-f694-4780-9d03-1938b4713629)

### History
* :broken_heart: [Build #220620](https://buildkite.com/elastic/kibana-pull-request/builds/220620) failed b83a88c70440bd629e76951ddf7e4a71a1d2950e
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222026","number":222026,"commit":"3dba3538ee1aded3c384bed3475fb56096a1ebea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":222026}
buildkite-pr-comment-->

jpdjere committed:
Added logic change for multi-line string algo

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222518)
* Commit: a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d

### Failed CI Steps
* [Jest Tests #12](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee)
* [Jest Tests #12](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f6c-48ff-9546-86486378181b) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if all three versions are different - scenario ABC returns a computated merged version without a conflict if 3 way merge is possible
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b8-4635-983d-e48f50ff4312) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if all three versions are different - scenario ABC returns a computated merged version without a conflict if 3 way merge is possible
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f73-4a36-a737-95a005c00221) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if all three versions are different - scenario ABC returns the current_version with a conflict if 3 way merge is not possible
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b3-4c33-9116-582bac3eaf55) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if all three versions are different - scenario ABC returns the current_version with a conflict if 3 way merge is not possible
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f73-44ac-a1a6-f7546b0fbec4) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if base_version is missing returns current_version as merged output if current_version and target_version are the same - scenario -AA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b4-48f5-977f-3b5b9a98cfff) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if base_version is missing returns current_version as merged output if current_version and target_version are the same - scenario -AA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f73-4f11-bbf8-9f4f21d81468) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if base_version is missing returns target_version as merged output if current_version and target_version are different - scenario -AB
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b7-49a7-b8c4-9bf61326bab5) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm if base_version is missing returns target_version as merged output if current_version and target_version are different - scenario -AB
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f70-40db-a810-4b6cc4abc36c) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if current version is different but it matches the update - scenario ABB
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b5-45de-bd6e-dee83c6b2c77) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if current version is different but it matches the update - scenario ABB
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f6d-4d15-9e8c-0640266349ed) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if current_version is different and there is no update - scenario ABA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b6-49ee-abe8-f9888ae5794d) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if current_version is different and there is no update - scenario ABA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f70-4863-bc2c-95b06b69952a) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if there is no update - scenario AAA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b3-4d3a-a50e-70319adf5b49) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns current_version as merged output if there is no update - scenario AAA
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca06-7d5a-4dcf-999b-a99588fd43ee) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca06-7d5a-4dcf-999b-a99588fd43ee/artifacts/0190ca26-6f6f-48bd-9ee1-1be58e12f207) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns target_version as merged output if current_version is the same and there is an update - scenario AAB
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222518#0190ca26-9611-4abb-9468-8b3807932a75) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222518/jobs/0190ca26-9611-4abb-9468-8b3807932a75/artifacts/0190ca33-24b1-43ad-a8fa-585d70099732) Jest Tests #<span></span>12 / multiLineStringDiffAlgorithm returns target_version as merged output if current_version is the same and there is an update - scenario AAB

### History
* :broken_heart: [Build #222026](https://buildkite.com/elastic/kibana-pull-request/builds/222026) failed 3dba3538ee1aded3c384bed3475fb56096a1ebea
* :broken_heart: [Build #220620](https://buildkite.com/elastic/kibana-pull-request/builds/220620) failed b83a88c70440bd629e76951ddf7e4a71a1d2950e
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222518","number":222518,"commit":"a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222026","number":222026,"commit":"3dba3538ee1aded3c384bed3475fb56096a1ebea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":222518}
buildkite-pr-comment-->

jpdjere committed:
Add has_base_version prop

jpdjere committed:
Tests for multi-line strings and missing version

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222570)
* Commit: 8f78e699825a7a7e99dd9dce2abae2b43497215e

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/222570#0190ca88-33a5-422c-ac76-a17026a9ae3b)

### History
* :broken_heart: [Build #222518](https://buildkite.com/elastic/kibana-pull-request/builds/222518) failed a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d
* :broken_heart: [Build #222026](https://buildkite.com/elastic/kibana-pull-request/builds/222026) failed 3dba3538ee1aded3c384bed3475fb56096a1ebea
* :broken_heart: [Build #220620](https://buildkite.com/elastic/kibana-pull-request/builds/220620) failed b83a88c70440bd629e76951ddf7e4a71a1d2950e
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222570","number":222570,"commit":"8f78e699825a7a7e99dd9dce2abae2b43497215e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222518","number":222518,"commit":"a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222026","number":222026,"commit":"3dba3538ee1aded3c384bed3475fb56096a1ebea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":222570}
buildkite-pr-comment-->

jpdjere committed:
Update integration tests

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222598)
* Commit: dab9bd3c278d5ca568a3658f2ab211548cdddf47

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/222598#0190cb09-cfc9-44c9-b4e4-5706f7427545)

### History
* :broken_heart: [Build #222570](https://buildkite.com/elastic/kibana-pull-request/builds/222570) failed 8f78e699825a7a7e99dd9dce2abae2b43497215e
* :broken_heart: [Build #222518](https://buildkite.com/elastic/kibana-pull-request/builds/222518) failed a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d
* :broken_heart: [Build #222026](https://buildkite.com/elastic/kibana-pull-request/builds/222026) failed 3dba3538ee1aded3c384bed3475fb56096a1ebea
* :broken_heart: [Build #220620](https://buildkite.com/elastic/kibana-pull-request/builds/220620) failed b83a88c70440bd629e76951ddf7e4a71a1d2950e
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222598","number":222598,"commit":"dab9bd3c278d5ca568a3658f2ab211548cdddf47"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222570","number":222570,"commit":"8f78e699825a7a7e99dd9dce2abae2b43497215e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222518","number":222518,"commit":"a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222026","number":222026,"commit":"3dba3538ee1aded3c384bed3475fb56096a1ebea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220383","number":220383,"commit":"7afd8975058375d103b204c8c1bab0725200437b"}],"number":222598}
buildkite-pr-comment-->

jpdjere committed:
Lint

jpdjere committed:
Update comment

nikitaindik commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff.ts:
```diff
@@ -98,6 +104,13 @@ export interface ThreeWayDiff<TValue> extends ThreeVersionsOf<TValue> {
+   * - true: the base version of the field was found and is either defined or undefined
+   * - false: the base version of the field was not found
+   */
+  has_base_version: boolean;
```
GitHub doesn't allow to post a comment on line 36, so posting it here.

```ts
export interface ThreeVersionsOf<TValue> {
  base_version: TValue | MissingVersion;
  ...
```

What are your thoughts on removing the `MissingVersion` type from `ThreeVersionsOf` and just relying on newly added `has_base_version` boolean?

Pros of removing:
 - A bit type safer because FE uses the same `ThreeWayDiff` type. And `ThreeWayDiff` says that `base_version` is `TValue | MissingVersion`, but on FE it'll actually be `TValue | undefined`.
 - One way to check if base version is missing (`has_base_version`) on backend vs two ways (`has_base_version` and `baseVersion === MissingValue`).

Pros of leaving as is:
 - We won't need to update the code in a bunch of places :)

jpdjere replied:
Not sure I get your proposal: what's your idea for the type? `base_version: TValue`?

In the backend we still need to distinguish the case where the base_version field exists and is undefined vs the case where the base_version does not exist. And we need to do certain checks such as:

https://github.com/elastic/kibana/blob/9faf2f6ebf06941445a05e2a9b63ea374cbcb7f8/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/calculate_rule_fields_diff.ts#L61

I.e: if we remove the possibility of `base_version` being MissingVersion, how we differentiate both cases?

nikitaindik replied:
I see why this is confusing. We use `MissingValue` to indicate one of two things:
 - base_version of a rule is missing
 - field value is missing because the base_version of a rule is missing (my idea was to remove `MissingValue` for this use case)

In the example you referred to (`calculate_rule_fields_diff.ts:61`) `MissingVersion` means "base_version of a rule is missing". I think it's okay and should stay as is.

But we could probably kick the `MissingVersion` out of `ThreeWayDiff`, where it means "field value is missing because the base_version of a rule is missing". We'll just have
```ts
interface ThreeWayDiff<TValue> {
 ...
 has_base_version: boolean;
 base_version?: TValue;
 ...
}
```

nikitaindik replied:
I just understood that then we can't change it in `ThreeVersionsOf` because it's used for both use cases. Then we can just put these fields directly ThreeWayDiff type definition without using `extends ThreeVersionsOf`.

Smth like:
```ts
interface ThreeWayDiff<TValue> {
  base_version?: TValue;
  current_version: TValue;
  target_version: TValue;
  // ... all the other ThreeWayDiff fields
}
``

Hope it makes sense

banderror replied:
@jpdjere `MissingVersion` is a `Symbol` and it is not serialized into JSON when returned from the API. This means that we should stop relying on it and should remove from the API contract. Both `ThreeWayDiff<TValue>` and `ThreeVersionsOf<TValue>` types is a part of the API contract, so we should refactor them in a way so they don't depend on `MissingVersion` anymore. One way to do it would be:

banderror replied:
And we probably should remove the `MissingVersion` constant from the codebase completely, to not have two ways of defining a missing base version.

banderror replied:
Another way to do it would be to remove inheritance from `ThreeWayDiff<TValue> extends ThreeVersionsOf<TValue>`:

```ts
export interface ThreeVersionsOf<TValue> {
  base_version: TValue | MissingVersion;
  current_version: TValue;
  target_version: TValue;
}
```

```ts
export interface ThreeWayDiff<TValue> {
  has_base_version: boolean;
  base_version: TValue | undefined;
  current_version: TValue;
  target_version: TValue;
  merged_version: TValue;
  diff_outcome: ThreeWayDiffOutcome;
  merge_outcome: ThreeWayMergeOutcome;
  has_update: boolean;
  conflict: ThreeWayDiffConflict;
}
```

If we do it, `ThreeVersionsOf<TValue>` will be removed from the API contract which means we could continue using `MissingVersion` in it.

dplumlee replied:
I think, from a diff writing perspective, it could be easier without the `MissingVersion` symbol in there just for typing sake but overall the logic would remain the same. Given the way we calculate `MissingVersion` [here](https://github.com/elastic/kibana/blob/2c8b2ff4a55756a345b0124f2beaef7a5ce849c9/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculate_rule_diff.ts#L73), I think it'd be a pretty easy removal if we add that extra `has_base_version` field

jpdjere replied:
Removed inheritance of `ThreeWayDiff`. Thanks for the proposals 👍

jpdjere committed:
Remove file

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222632)
* Commit: 2c8741b31c328ba1607326cc047adf2b452aceed

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/222632#0190cb59-5ff9-4dbc-986a-a5ee3fdec9f7)

### History
* :broken_heart: [Build #222598](https://buildkite.com/elastic/kibana-pull-request/builds/222598) failed dab9bd3c278d5ca568a3658f2ab211548cdddf47
* :broken_heart: [Build #222570](https://buildkite.com/elastic/kibana-pull-request/builds/222570) failed 8f78e699825a7a7e99dd9dce2abae2b43497215e
* :broken_heart: [Build #222518](https://buildkite.com/elastic/kibana-pull-request/builds/222518) failed a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d
* :broken_heart: [Build #222026](https://buildkite.com/elastic/kibana-pull-request/builds/222026) failed 3dba3538ee1aded3c384bed3475fb56096a1ebea
* :broken_heart: [Build #220620](https://buildkite.com/elastic/kibana-pull-request/builds/220620) failed b83a88c70440bd629e76951ddf7e4a71a1d2950e
* :broken_heart: [Build #220383](https://buildkite.com/elastic/kibana-pull-request/builds/220383) failed 7afd8975058375d103b204c8c1bab0725200437b

cc @jpdjere

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222632","number":222632,"commit":"2c8741b31c328ba1607326cc047adf2b452aceed"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222598","number":222598,"commit":"dab9bd3c278d5ca568a3658f2ab211548cdddf47"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222570","number":222570,"commit":"8f78e699825a7a7e99dd9dce2abae2b43497215e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222518","number":222518,"commit":"a5c1c99aa0aa31c0d95fcd3bc7884120e79d2a9d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222026","number":222026,"commit":"3dba3538ee1aded3c384bed3475fb56096a1ebea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220620","number":220620,"commit":"b83a88c70440bd629e76951ddf7e4a71a1d2950e"}],"number":222632}
buildkite-pr-comment-->

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/rule_diff/rule_diff.ts:
```diff
@@ -63,6 +63,10 @@ export type RuleFieldsDiff = CommonFieldsDiff &
 export interface FullRuleDiff {
   fields: RuleFieldsDiff;
   has_conflict: boolean;
+  has_non_solvable_conflict: boolean;
+  num_fields_with_updates: number;
+  num_fields_with_conflicts: number;
+  num_fields_with_non_solvable_conflicts: number;
```
If we have `num_fields_with_conflicts` and `num_fields_with_non_solvable_conflicts`, do we still need `has_conflict` and `has_non_solvable_conflict`?

jpdjere replied:
These two fields simplified the calculation of `num_rules_with_conflicts` and `num_rules_with_non_solvable_conflicts` in `calculateRuleStats` in `x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_upgrade/review_rule_upgrade_route.ts`.

But, yeah, agree that its cleaner to remove them from the types and calculate these two in another way. I rewrote the `calculateRuleStats` method.

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/rule_diff/rule_diff.ts:
```diff
@@ -72,6 +76,10 @@ export interface FullRuleDiff {
 export interface PartialRuleDiff {
   fields: Partial<RuleFieldsDiff>;
   has_conflict: boolean;
+  has_non_solvable_conflict: boolean;
+  num_fields_with_updates: number;
+  num_fields_with_conflicts: number;
+  num_fields_with_non_solvable_conflicts: number;
```
Nit:

```ts
interface BaseRuleDiff {
  has_conflict: boolean;
  has_non_solvable_conflict: boolean;
  num_fields_with_updates: number;
  num_fields_with_conflicts: number;
  num_fields_with_non_solvable_conflicts: number;
}

/**
 * Full rule diff contains diffs for all the top-level rule fields.
 * Even if there's no change at all to a given field, its diff will be included in this object.
 * This diff can be useful for internal server-side calculations or debugging.
 * Note that this is a pretty large object so returning it from the API might be undesirable.
 */
export interface FullRuleDiff extends BaseRuleDiff {
  fields: RuleFieldsDiff;
}

/**
 * Partial rule diff contains diffs only for those rule fields that have some changes to them.
 * This diff can be useful for returning info from REST API endpoints because its size is tolerable.
 */
export interface PartialRuleDiff extends BaseRuleDiff {
  fields: Partial<RuleFieldsDiff>;
}
```

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff.ts:
```diff
@@ -19,6 +19,12 @@ import type { ThreeWayMergeOutcome } from './three_way_merge_outcome';
 export const MissingVersion = Symbol('MissingVersion');
 export type MissingVersion = typeof MissingVersion;
 
+export enum ThreeWayDiffConflictResolutionResult {
+  NO_CONFLICT = 'NO_CONFLICT',
+  SOLVABLE_CONFLICT = 'SOLVABLE_CONFLICT',
+  NON_SOLVABLE_CONFLICT = 'NON_SOLVABLE_CONFLICT',
+}
```
For the sake of brevity, could we please do this:

```suggestion
export enum ThreeWayDiffConflict {
  NONE = 'NONE',
  SOLVABLE = 'SOLVABLE',
  NON_SOLVABLE = 'NON_SOLVABLE',
}
```

Also:

- I think it deserves its own file in consistency with `ThreeWayDiffOutcome` and `ThreeWayMergeOutcome`.
- Please add JSDoc comments for the whole enum and its values. It would be great to refer to the corresponding sections of the RFC.

jpdjere replied:
This was initially so and I expanded for clarity after getting some [feedback.](https://github.com/elastic/kibana/pull/187770#discussion_r1679916323)

But I'll extract it to its own file and add JSDoc comments explaining this and that should be enough to make the enum clear enough to anyone using it.

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/review_rule_upgrade/review_rule_upgrade_route.ts:
```diff
@@ -21,6 +21,12 @@ export interface RuleUpgradeStatsForReview {
   /** Number of installed prebuilt rules available for upgrade (stock + customized) */
   num_rules_to_upgrade_total: number;
 
+  /** Number of installed prebuilt rules with upgrade conflicts (SOLVABLE or NON_SOLVALBE) */
+  num_rules_with_conflicts: number;
+
+  /** Number of installed prebuilt rules with NON_SOLVABLE upgrade conflicts */
+  num_rules_with_non_solvable_conflicts: number;
```
Why do we need this information in the `upgrade/_review` response?

jpdjere replied:
The use case I have in my mind for these two props is the [Add ability to bulk update all rules that don't have conflicts #180589](https://github.com/elastic/kibana/issues/180589).

We'll want a bulk update button that opens a modal with text that looks like:

`${num_rules_to_upgrade_total} Elastic rules have updates. ${num_rules_to_upgrade_total - num_rules_with_non_solvable_conflicts} will be updated automatically, but ${num_rules_with_non_solvable_conflicts} have conflicts that will need manual resolution`.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculate_rule_diff.ts:
```diff
@@ -75,14 +77,45 @@ export const calculateRuleDiff = (args: RuleVersions): CalculateRuleDiffResult =
 
-  const hasAnyFieldConflict = Object.values<ThreeWayDiff<unknown>>(fieldsDiff).some(
-    (fieldDiff) => fieldDiff.has_conflict
+  const {
+    numberFieldsWithUpdates,
+    numberFieldsWithConflicts,
+    numberFieldsWithNonSolvableConflicts,
+  } = Object.values(fieldsDiff).reduce<{
+    numberFieldsWithUpdates: number;
+    numberFieldsWithConflicts: number;
+    numberFieldsWithNonSolvableConflicts: number;
+  }>(
```
This calculation occupies a significant portion of the body of the `calculateRuleDiff` function, while not being the most significant part of it semantically. Can we extract it to another function?

jpdjere replied:
Extracted out

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculate_rule_diff.ts:
```diff
@@ -75,14 +77,45 @@ export const calculateRuleDiff = (args: RuleVersions): CalculateRuleDiffResult =
+    numberFieldsWithConflicts: number;
+    numberFieldsWithNonSolvableConflicts: number;
+  }>(
+    (counts, fieldDiff) => {
+      if (fieldDiff.has_update) {
+        counts.numberFieldsWithUpdates += 1;
+      }
```
`fieldDiff` has a type of `any`, and the code of this arrow function is not type-safe.

To fix this, we can return back the generic parameter specified for `Object.values`:

```ts
Object.values<ThreeWayDiff<unknown>>(fieldsDiff).reduce<...
```

jpdjere replied:
Added back

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.ts:
```diff
@@ -112,3 +114,13 @@ const mergeVersions = ({
   }
 };
+
+const determineConflictType = (mergeOutcome: ThreeWayMergeOutcome) => {
+  if (mergeOutcome === ThreeWayMergeOutcome.Conflict) {
+    return ThreeWayDiffConflictResolutionResult.NON_SOLVABLE_CONFLICT;
+  }
+  if (mergeOutcome === ThreeWayMergeOutcome.Merged) {
+    return ThreeWayDiffConflictResolutionResult.SOLVABLE_CONFLICT;
+  }
+  return ThreeWayDiffConflictResolutionResult.NO_CONFLICT;
+};
```
Two suggestions:

1. I think we don't need a `ThreeWayMergeOutcome.Conflict` enum value anymore, because now we have a separate enum for a conflict.
2. The value of `ThreeWayDiffConflictResolutionResult` should be determined in the `mergeVersions` function where we decide  how to handle various diff outcomes, including the ABC one (`ThreeWayDiffOutcome.CustomizedValueCanUpdate:`).

jpdjere replied:
I'll remove it for now, but we can reconsider it in the future if the need arises.

jpdjere replied:
@banderror I implemented your suggestions, and I indeed think it looks cleaner. I'm thinking now if we should go one step further and get rid of the special logic to handle missing versions, in the [concrete diff algorithms](https://github.com/elastic/kibana/pull/187770/files#diff-ffefa779bbd4ca132f68af89e01b1d772507796db082ef10a820a555ddfcfa02R90) by extending the `ThreeWayDiffOutcome` enum to include the [two special scenarios of the compromise solution](https://github.com/elastic/kibana/pull/184889#discussion_r1636421293):

```
 */
export enum ThreeWayDiffOutcome {
  /** Missing  base, the value hasn't changed in the target version. */
  MissingBaseNoUpdate = 'BASE=-, CURRENT=A, TARGET=A',

  /** Missing  base, the value changed in the target version. */
  MissingBaseWithUpdate = 'BASE=-, CURRENT=A, TARGET=B',

  // ... and the existing scenarios
}

dplumlee replied:
I actually think this would improve overall code context as right now, we just treat these edge cases as if they were something different (e.g. `-AA` as `AAA`). This would allow us to have better control over the specific situations without needing to have extra logic parsing the outcome anywhere downstream from the diffing logic.

banderror replied:
@jpdjere Please go ahead, it's a great idea 👍

banderror reviewed (commented):
Overall the PR looks good, but let's spend more time on its quality before merging it.

jpdjere committed:
Address feedback

jpdjere committed:
Replaced enum

jpdjere committed:
Merge branch 'main' into extend-upgrade-review-endpoint

jpdjere committed:
Removed redudant API response props

jpdjere committed:
Handled missing base version scenarios

jpdjere committed:
Updated integration tests

jpdjere committed:
Remove changes to rfc

jpdjere commented:
@banderror @nikitaindik @dplumlee 

Thanks for the review. I've addressed your feedback, and included in the changes the handling of scenarios with a missing base version, with the "compromise" solution we had decided [here.](https://github.com/elastic/kibana/pull/184889#discussion_r1636421293)

Please take a look when you can.

jpdjere committed:
Added link to RFC

jpdjere committed:
Remove consolelog

jpdjere committed:
remove index

jpdjere committed:
Fix

nikitaindik reviewed (approved):
Thanks for the updates to PR @jpdjere! I have only reviewed changes related to `MissingVersion` - they look good to me!

dplumlee commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.test.ts:
```diff
@@ -144,10 +138,11 @@ describe('multiLineStringDiffAlgorithm', () => {
       expect(result).toEqual(
         expect.objectContaining({
+          has_base_version: false,
+          base_version: undefined,
```
👍

dplumlee commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.ts:
```diff
@@ -34,77 +34,91 @@ export const multiLineStringDiffAlgorithm = (
-    case ThreeWayDiffOutcome.StockValueNoUpdate:
-    case ThreeWayDiffOutcome.CustomizedValueNoUpdate:
-    case ThreeWayDiffOutcome.CustomizedValueSameUpdate: {
+    case ThreeWayDiffOutcome.StockValueNoUpdate: // Scenarios AAA and -AA
+    case ThreeWayDiffOutcome.CustomizedValueNoUpdate: // Scenario ABA
+    case ThreeWayDiffOutcome.CustomizedValueSameUpdate: // Scenario ABB
```
Nice idea to put these scenarios in here, I think I'll incorporate this into some other diff files as well

jpdjere replied:
@dplumlee I ended up removing these, since now each scenario has its own `ThreeWayDiffOutcome` and there's no case like `// Scenarios AAA and -AA` anymore.

dplumlee reviewed (approved):
The new changes look great to me @jpdjere, thanks for incorporating all this into the existing diff work! I do like the idea you mentioned [here](https://github.com/elastic/kibana/pull/187770#discussion_r1686761793), not sure if there's a reason we didn't incorporate this before other than not really having any special logic needs for those cases with the initial implementation

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_merge_outcome.ts:
```diff
@@ -1,26 +0,0 @@
- *   - current version
- *   - target version
- */
-export enum ThreeWayMergeOutcome {
-  /** Took current version and returned as the merged one. */
-  Current = 'CURRENT',
-
-  /** Took target version and returned as the merged one. */
-  Target = 'TARGET',
-
-  /** Merged three versions successfully into a new one. */
-  Merged = 'MERGED',
```
This is probably too bold move. My ask was to remove the `Conflict` value, but not the whole property from the `ThreeWayDiff` object. Please add the `ThreeWayMergeOutcome` back.

Although it's not needed for implementing diff calculations or any UI, it was supposed to be used as a property that might be helpful for debugging and troubleshooting of diff calculations or the whole upgrade workflow in general. I expect it to be especially helpful for understanding the diffing of complex objects, such as `related_integrations` and `threat` arrays.

jpdjere replied:
Ok, I'm adding this back. 

Just to be 100% sure I understand your initial idea: with the new `ThreeWayDiffConflict` enum, we don't need the `ThreeWayMergeOutcome.conflict` value anymore, because `ThreeWayMergeOutcome` should only be able to return the following values:

- `Merged`: only for `SOLVABLE` conflicts - a conflict existed between the three version but our algorithm could create a merged solution out of the three
- `Current` and `Target`: two options:
    - There was `NONE` conflict, and the algorithm decided that the returned value should be either `Current` or `Target` (this depends on the algorithm and the specific scenario)
    - There was `NON_SOLVABLE` conflict, no merged version was possible, so we are returning either `Current` or `Target` (this also depends on the algorithm and the specific scenario, but in almost all cases it will be `Current`)

Are we aligned here?

banderror replied:
@jpdjere Your understanding is exactly right! Thank you.

In other words:

- In trivial cases, such as `AAA`, `ABA`, `AAB`, `ABB`, we return:
  - `ThreeWayDiffConflict`: `NONE`
  - `ThreeWayMergeOutcome`: `Current` or `Target`, depending on the case and algorithm applied to the field
- In the `ABC` cases when it's not possible to auto-merge or we think we should not, we return:
  - `ThreeWayDiffConflict`: `NON_SOLVABLE`
  - `ThreeWayMergeOutcome`: `Current` or `Target`, depending on the algorithm (`Current` in most cases)
- In the `ABC` cases when it's possible to auto-merge and we do it, we return:
  - `ThreeWayDiffConflict`: `SOLVABLE`
  - `ThreeWayMergeOutcome`: `Merged`, even if the resulting auto-merged version that the algorithm outputs turns out to be equal to the current or target versions

jpdjere committed:
Readded ThreeWayMergeOutcome enum and added missing base version scenarios to ThreeWayDiffOutcome enum

jpdjere committed:
Simplified single line strings integration tests

jpdjere committed:
Simplified integration tests

jpdjere committed:
Merge branch 'main' into extend-upgrade-review-endpoint

jpdjere commented:
@banderror Reintroduced `ThreeWayMergeOutcome` enum as we discussed.

@dplumlee While introducing/modifying the integration tests for scenarios `MissingBaseNoUpdate` and `MissingBaseCanUpdate`, I realised that doing assertions on the whole `reviewResponse.rules[0].diff.fields` object was confusing, since `version` is always present and we need to describe its behaviour in all unrelated tests. So I went ahead and simplified them, now doing assertions only on the field that should/shouldn't be part of the `diff` object - so that if the scenario excludes the field from the diff object, then the assertion is just `expect(reviewResponse.rules[0].diff.fields.name).toBeUndefined()`.

So now the focus of each test is exclusively on the field that corresponds to the algorithm type.

jpdjere committed:
Merge branch 'main' into extend-upgrade-review-endpoint

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/scalar_array_diff_algorithm.ts:
```diff
@@ -36,75 +37,81 @@ export const scalarArrayDiffAlgorithm = <TValue>(
-      }
-
-      const addedCurrent = difference(dedupedCurrentVersion, dedupedBaseVersion);
+      const addedCurrent = difference(dedupedCurrentVersion, dedupedBaseVersion as TValue[]);
```
We always try to avoid `as` casting in production code, can we try to rewrite this algorithm in a type-safe way?

jpdjere replied:
Fixed

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/scalar_array_diff_algorithm.ts:
```diff
@@ -36,75 +37,81 @@ export const scalarArrayDiffAlgorithm = <TValue>(
       const removedCurrent = difference(dedupedBaseVersion, dedupedCurrentVersion);
 
-      const addedTarget = difference(dedupedTargetVersion, dedupedBaseVersion);
+      const addedTarget = difference(dedupedTargetVersion, dedupedBaseVersion as TValue[]);
```
Same here

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.ts:
```diff
@@ -34,80 +35,102 @@ export const multiLineStringDiffAlgorithm = (
-      const mergedVersion = merge(currentVersion, baseVersion, targetVersion, {
+      // TS does not realize that in ABC scenario, baseVersion cannot be missing
+      // Missing baseVersion scenarios were handled as -AA and -AB.
+      const mergedVersion = merge(currentVersion, baseVersion as string, targetVersion, {
```
Same here

banderror reviewed (approved):
Thanks @jpdjere, I left a couple more comments. Approving in advance 👍

jpdjere committed:
Remove type casting

jpdjere committed:
Merge branch 'main' into extend-upgrade-review-endpoint
